crates/ruff/src/importer/insertion.rs

```diff
@@ -34,86 +100,174 @@ impl Insertion {
                 locator.full_line_end(location),
                 stylist.line_ending().as_str(),
             )
         }
     }
 
-    /// Create an [`Insertion`] to insert (e.g.) an import statement at the "top" of a given file,
-    /// along with a prefix and suffix to use for the insertion.
+    /// Create an [`Insertion`] to insert (e.g.) an import statement at the start of a given
+    /// block, along with a prefix and suffix to use for the insertion.
     ///
     /// For example, given the following code:
     ///
     /// ```python
-    /// """Hello, world!"""
-    ///
-    /// import os
+    /// if TYPE_CHECKING:
+    ///     import os
     /// ```
     ///
     /// The insertion returned will begin at the start of the `import os` statement, and will
-    /// include a trailing newline suffix.
-    pub(super) fn start_of_file(body: &[Stmt], locator: &Locator, stylist: &Stylist) -> Insertion {
-        // Skip over any docstrings.
-        let mut location = if let Some(location) = match_docstring_end(body) {
-            // If the first token after the docstring is a semicolon, insert after the semicolon as an
-            // inline statement;
-            let first_token = lexer::lex_starts_at(locator.after(location), Mode::Module, location)
-                .flatten()
-                .next();
-            if let Some((Tok::Semi, range)) = first_token {
-                return Insertion::new(" ", range.end(), ";");
-            }
-
-            // Otherwise, advance to the next row.
-            locator.full_line_end(location)
-        } else {
-            TextSize::default()
-        };
+    /// include a trailing newline.
+    ///
+    /// The block itself is assumed to be at the top-level of the module.
+    pub(super) fn start_of_block(
+        mut location: TextSize,
+        locator: &Locator<'a>,
+        stylist: &Stylist,
+    ) -> Insertion<'a> {
+        enum Awaiting {
+            Colon(u32),
+            Newline,
+            Indent,
+        }
 
-        // Skip over any comments and empty lines.
+        let mut state = Awaiting::Colon(0);
         for (tok, range) in
             lexer::lex_starts_at(locator.after(location), Mode::Module, location).flatten()
         {
-            if matches!(tok, Tok::Comment(..) | Tok::Newline) {
-                location = locator.full_line_end(range.end());
-            } else {
-                break;
+            match state {
+                // Iterate until we find the colon indicating the start of the block body.
+                Awaiting::Colon(depth) => match tok {
+                    Tok::Colon if depth == 0 => {
+                        state = Awaiting::Newline;
+                    }
+                    Tok::Lpar | Tok::Lbrace | Tok::Lsqb => {
+                        state = Awaiting::Colon(depth + 1);
+                    }
+                    Tok::Rpar | Tok::Rbrace | Tok::Rsqb => {
+                        state = Awaiting::Colon(depth - 1);
```

Comment:
Nit: To avoid panicking if the parentheses are unbalanced. 

```suggestion
                        state = Awaiting::Colon(depth.saturdating_sub(1));
```